gobblin-data-management/src/main/java/gobblin/data/management/retention/dataset/DatasetBase.java

```diff
@@ -0,0 +1,215 @@
+  public static final boolean SKIP_TRASH_DEFAULT = false;
+  public static final String DELETE_EMPTY_DIRECTORIES_KEY = CONFIGURATION_KEY_PREFIX + "delete.empty.directories";
+  public static final boolean DELETE_EMPTY_DIRECTORIES_DEFAULT = true;
+
+  protected final FileSystem fs;
+  protected final Trash trash;
+  protected final boolean simulate;
+  protected final boolean skipTrash;
+  protected final boolean deleteEmptyDirectories;
+
+  /**
+   * Get {@link gobblin.data.management.retention.version.finder.VersionFinder} to use.
+   */
+  public abstract VersionFinder<? extends T> getVersionFinder();
+
+  /**
+   * Get {@link gobblin.data.management.retention.policy.RetentionPolicy} to use.
+   * @return
+   */
+  public abstract RetentionPolicy<T> getRetentionPolicy();
+
+  public DatasetBase(FileSystem fs, Props props) throws IOException {
+    this.fs = fs;
+
+    this.simulate = props.getBoolean(SIMULATE_KEY, SIMULATE_DEFAULT);
+    this.skipTrash = props.getBoolean(SKIP_TRASH_KEY, SKIP_TRASH_DEFAULT);
+    this.deleteEmptyDirectories = props.getBoolean(DELETE_EMPTY_DIRECTORIES_KEY, DELETE_EMPTY_DIRECTORIES_DEFAULT);
+
+    if(this.simulate || this.skipTrash) {
+      this.trash = null;
+    } else {
+      this.trash = new Trash(fs, props);
+    }
+  }
+
+  public DatasetBase(FileSystem fs, Trash trash, boolean simulate, boolean skipTrash, boolean deleteEmptyDirectories) {
+    this.fs = fs;
+    this.trash = trash;
+    this.simulate = simulate;
+    this.skipTrash = skipTrash;
+    this.deleteEmptyDirectories = deleteEmptyDirectories;
+  }
+
+  /**
+   * Perform the cleanup of old / deprecated dataset versions. See {@link gobblin.data.management.retention.DatasetCleaner}
+   * javadoc for more information.
+   * @throws java.io.IOException
+   */
+  @Override
+  public void clean() throws IOException {
+
+    RetentionPolicy<T> retentionPolicy = getRetentionPolicy();
+    VersionFinder<? extends T> versionFinder = getVersionFinder();
+
+    if(!retentionPolicy.versionClass().isAssignableFrom(versionFinder.versionClass())) {
+      throw new IOException("Incompatible dataset version classes.");
+    }
+
+    LOGGER.info("Cleaning dataset " + this);
+
+    List<T> versions = Lists.newArrayList(getVersionFinder().findDatasetVersions(this));
+    Collections.sort(versions, Collections.reverseOrder());
+
+    List<T> deletableVersions =
+        getRetentionPolicy().preserveDeletableVersions(versions);
+
+    Set<Path> possiblyEmptyDirectories = new HashSet<Path>();
+
+    for(DatasetVersion versionToDelete : deletableVersions) {
+      LOGGER.info("Deleting dataset version " + versionToDelete);
+
+      Set<Path> pathsToDelete = versionToDelete.getPathsToDelete();
+      LOGGER.info("Deleting paths: " + Arrays.toString(pathsToDelete.toArray()));
+
+      boolean deletedAllPaths = true;
+
+      for(Path path: pathsToDelete) {
+        if (!this.simulate) {
```

Comment:
We should log the directories that would have been deleted in the simulate mode.